util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,238 @@
+#!/usr/bin/env python3
+# Copyright (c) 2023 The Regents of the University of California
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met: redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer;
+# redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution;
+# neither the name of the copyright holders nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+This is a script to get all resources from the database and save them to
+a json file and download all resources to a directory. The resources in
+the resources.json file that is generated will be updated to point to the
+local resources instead of the database resources. This script is used to
+create the offline database. To use the resources.json with gem5 you need
+to update the gem5 config to point to the resources.json file.
+"""
+
+import argparse
+import json
+from urllib import request, parse
+import pathlib
+from typing import List
+import hashlib
+from tqdm import tqdm
+
+
+def get_token(auth_url: str, api_key: str) -> str:
+    """
+    This function gets the token from the database using the api key
+    :param auth_url: Authentication url to use Atlas data API
+    :param api_key: API key to access the database
+    :return: Token to access the database
+    """
+
+    data = {"key": api_key}
+    data = json.dumps(data).encode("utf-8")
+
+    req = request.Request(
+        auth_url, data=data, headers={"content-type": "application/json"}
+    )
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(f"Failed to get token: {e}")
```

Comment:
Nitpik:

```suggestion
        raise Exception("Failed to obtain token via url request. "
                        f"Exception thrown:\n{e}")
```

1. I prefer printing the exception received on a new line.
2. Reading this, at least in code makes me think the token request is going to proceed the `:`. `Failed to get token: APT_TOKEN_1`. This is probably just me, but my phrasing increases the clarity by a tiny amount..